File: darwin/pro_video_editor/Sources/pro_video_editor/src/shared/features/audio/models/AudioMergeConfig.swift
import Foundation

/// A single trimmed audio window that participates in an audio merge.
struct AudioMergeSegmentConfig {
  /// Absolute file path to the source video/audio for this segment.
  let inputPath: String

  /// Window start in the source timeline, in microseconds.
  let startUs: Int64

  /// Window end in the source timeline, in microseconds (`> startUs`).
  let endUs: Int64

  /// Playback speed multiplier applied after trimming (`> 0`, pitch preserved).
  let speed: Double

  /// The played length this segment contributes to the output, in microseconds
  /// (`(endUs - startUs) / speed`).
  var outputDurationUs: Double {
    Double(endUs - startUs) / speed
  }
}

/// Configuration model for merging several trimmed audio windows into one file.
///
/// Mirrors `AudioExtractConfig` but carries an ordered list of segments plus an
/// optional uniform output sample rate / channel count.
struct AudioMergeConfig {
  /// Unique task identifier for progress tracking and cancellation.
  let id: String

  /// Desired output format (`wav`, `aac`, `m4a`, `caf`, ...).
  let format: String

  /// Optional uniform output sample rate in Hz. `nil` = derive from the first
  /// audio-bearing segment.
  let sampleRate: Int?

  /// Optional uniform output channel count. `nil` = derive from the first
  /// audio-bearing segment.
  let channels: Int?

  /// Absolute path where the merged file is written.
  let outputPath: String

  /// Ordered segments to concatenate.
  let segments: [AudioMergeSegmentConfig]

  /// Whether the caller pinned an explicit uniform output format.
  var hasExplicitFormat: Bool { sampleRate != nil || channels != nil }

  /// Builds an `AudioMergeConfig` from Flutter method-call arguments.
  ///
  /// Returns `nil` if a required field is missing or a segment is invalid
  /// (`endUs <= startUs`, `speed <= 0`, empty list).
  static func fromArguments(_ arguments: [String: Any]?) -> AudioMergeConfig? {
    guard let args = arguments,
      let id = args["id"] as? String,
      let format = args["format"] as? String,
      let outputPath = args["outputPath"] as? String,
      let rawSegments = args["segments"] as? [[String: Any]],
      !rawSegments.isEmpty
    else {
      return nil
    }

    var segments: [AudioMergeSegmentConfig] = []
    for raw in rawSegments {
      guard let inputPath = raw["inputPath"] as? String,
        let startUs = (raw["startTime"] as? NSNumber)?.int64Value,
        let endUs = (raw["endTime"] as? NSNumber)?.int64Value
      else {
        return nil
      }
      let speed = (raw["speed"] as? NSNumber)?.doubleValue ?? 1.0
      if endUs <= startUs || speed <= 0 {
        return nil
      }
      segments.append(
        AudioMergeSegmentConfig(
          inputPath: inputPath, startUs: startUs, endUs: endUs, speed: speed))
    }

    let sampleRate = (args["sampleRate"] as? NSNumber)?.intValue
    let channels = (args["channels"] as? NSNumber)?.intValue

    return AudioMergeConfig(
      id: id,
      format: format,
      sampleRate: sampleRate,
      channels: channels,
      outputPath: outputPath,
      segments: segments)
  }

  /// Returns the output file extension for the configured format.
  func getOutputExtension() -> String {
    switch format.lowercased() {
    case "mp3": return "mp3"
    case "aac": return "m4a"
    case "m4a": return "m4a"
    case "caf": return "caf"
    case "wav": return "wav"
    default: return "m4a"
    }
  }

  /// Builds the equivalent single-clip `AudioExtractConfig` for a segment.
  ///
  /// Used for the single-segment fast path, which delegates to `ExtractAudio`
  /// to guarantee byte-for-byte parity with `extractAudioToFile`.
  func extractConfig(for segment: AudioMergeSegmentConfig) -> AudioExtractConfig {
    AudioExtractConfig(
      id: id,
      inputPath: segment.inputPath,
      fileExtension: "",
      format: format,
      startUs: segment.startUs,
      endUs: segment.endUs,
      speed: segment.speed,
      outputPath: outputPath)
  }
}

File: android/src/main/kotlin/ch/waio/pro_video_editor/src/features/audio/models/AudioMergeConfig.kt
package ch.waio.pro_video_editor.src.features.audio.models

import io.flutter.plugin.common.MethodCall

/**
 * A single trimmed audio window that participates in an audio merge.
 *
 * @property inputPath Absolute path to the source video/audio file.
 * @property startUs Window start in the source timeline, in microseconds.
 * @property endUs Window end in the source timeline, in microseconds (> startUs).
 * @property speed Playback speed multiplier applied after trimming (> 0).
 */
data class AudioMergeSegmentConfig(
    val inputPath: String,
    val startUs: Long,
    val endUs: Long,
    val speed: Float
)

/**
 * Configuration for merging several trimmed audio windows into one file.
 *
 * Mirrors [AudioExtractConfig] but carries an ordered list of segments plus an
 * optional uniform output sample rate / channel count.
 *
 * @property id Unique task identifier for progress tracking and cancellation.
 * @property format Output audio format (wav, aac, m4a, ...).
 * @property sampleRate Optional uniform output sample rate in Hz (null = derive
 *   from the first audio-bearing segment).
 * @property channels Optional uniform output channel count (null = derive from
 *   the first audio-bearing segment).
 * @property outputPath Absolute path where the merged file is written.
 * @property segments Ordered segments to concatenate.
 */
data class AudioMergeConfig(
    val id: String,
    val format: String,
    val sampleRate: Int?,
    val channels: Int?,
    val outputPath: String,
    val segments: List<AudioMergeSegmentConfig>
) {
    /** Whether the caller pinned an explicit uniform output format. */
    val hasExplicitFormat: Boolean get() = sampleRate != null || channels != null

    companion object {
        /**
         * Creates an [AudioMergeConfig] from a Flutter [MethodCall].
         *
         * @throws IllegalArgumentException if required parameters are missing or
         *   a segment is invalid (empty list, `endUs <= startUs`, `speed <= 0`).
         */
        fun fromMethodCall(call: MethodCall): AudioMergeConfig {
            val id = call.argument<String>("id")
                ?: throw IllegalArgumentException("id is required")
            val format = call.argument<String>("format") ?: "wav"
            val outputPath = call.argument<String>("outputPath")
                ?: throw IllegalArgumentException("outputPath is required")
            val sampleRate = call.argument<Number>("sampleRate")?.toInt()
            val channels = call.argument<Number>("channels")?.toInt()

            val rawSegments = call.argument<List<Map<String, Any?>>>("segments")
                ?: throw IllegalArgumentException("segments is required")
            if (rawSegments.isEmpty()) {
                throw IllegalArgumentException("segments must not be empty")
            }

            val segments = rawSegments.mapIndexed { index, raw ->
                val inputPath = raw["inputPath"] as? String
                    ?: throw IllegalArgumentException("segment[$index].inputPath is required")
                val startUs = (raw["startTime"] as? Number)?.toLong()
                    ?: throw IllegalArgumentException("segment[$index].startTime is required")
                val endUs = (raw["endTime"] as? Number)?.toLong()
                    ?: throw IllegalArgumentException("segment[$index].endTime is required")
                val speed = (raw["speed"] as? Number)?.toFloat() ?: 1.0f
                if (endUs <= startUs) {
                    throw IllegalArgumentException("segment[$index].endTime must be > startTime")
                }
                if (speed <= 0f) {
                    throw IllegalArgumentException("segment[$index].speed must be > 0")
                }
                AudioMergeSegmentConfig(
                    inputPath = inputPath,
                    startUs = startUs,
                    endUs = endUs,
                    speed = speed
                )
            }

            return AudioMergeConfig(
                id = id,
                format = format,
                sampleRate = sampleRate,
                channels = channels,
                outputPath = outputPath,
                segments = segments
            )
        }
    }

    /** Returns the file extension for the configured output format. */
    fun getExtension(): String = when (format.lowercase()) {
        "wav" -> "wav"
        "aac" -> "aac"
        "m4a" -> "m4a"
        "mp3" -> "mp3"
        "ogg" -> "ogg"
        else -> "m4a"
    }

    /** Builds the equivalent single-clip [AudioExtractConfig] for a segment. */
    fun extractConfig(segment: AudioMergeSegmentConfig): AudioExtractConfig = AudioExtractConfig(
        id = id,
        inputPath = segment.inputPath,
        format = format,
        startUs = segment.startUs,
        endUs = segment.endUs,
        speed = segment.speed,
        outputPath = outputPath
    )
}
